scalable-go-service/index.md

```diff
@@ -0,0 +1,322 @@
+#  Writing Scalable Go Services
+
+## Problem
+
+How do we show the customer a complete set of relevant hotel rates before they
+lose interest and close the tab? Despite trying various strategies to address
+this, the team will likely continue to face this challenge in the future as user
+expectations on speed and accuracy grow. The bottom line is, we need to load a
+lot of rates from many different vendors to show that we have as complete of an
+inventory as any other travel site. And we need to do this fast.
+
+Two of our main hotel vendors have announced major changes to the APIs we use to
+load hotel rates. We saw this as an opportunity to reevaluate and improve the
+backend services that interact with these APIs. To do this, we need to use
+different strategies to load rates based on the vendor and their API contracts.
+Most notably, some vendors allow us to get all available rates for multiple
+hotels in a single request while others only give data on an individual hotel
+basis.
+
+The latter case causes a lot of bottleneck for the shopping flow because even
+though the individual requests resolve fairly quickly, the time complexity grows
+linearly with the number of hotels. This would scale poorly as we increase the
+number of hotels we search. It also hinders user experience and confidence in
+the service if the user has to wait several seconds or even minutes for the
+search results to come back.
+
+## Concurrency and Go
+
+Since we have no influence on how our vendors' APIs are designed, it was clear
```

Comment:
This paragraph does not provide any new information. Should probably be mostly deleted / merged with the "why go" portion of the next paragraph.